Fix #1106: Addition of Work Manager for uploading logs

Sarthak2601 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #1106 - Implements work manager for uploading events and exceptions to Firebase in presence of network connection. 

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Sarthak2601 committed:
work manager start.

Sarthak2601 committed:
constraints + lint fixes.

Sarthak2601 committed:
work manager module + injection + onStart setup.

Sarthak2601 committed:
Merge branch 'develop' into oppia-work-manager

# Conflicts:
#	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt

Sarthak2601 committed:
lint fixes.

Sarthak2601 committed:
approach change.

vinitamurthi commented on domain/src/main/java/org/oppia/domain/oppialogger/OppiaLogUploadWorker.kt:
```diff
@@ -0,0 +1,81 @@
+      }
+      Result.success()
+    } catch (e: Exception) {
+      e.printStackTrace()
```
Please use the console logger here

Sarthak2601 replied:
Done.

rt4914 replied:
Resolving this now.

vinitamurthi commented on domain/src/main/java/org/oppia/domain/oppialogger/OppiaLogUploadWorker.kt:
```diff
@@ -0,0 +1,81 @@
+      }
+      Result.success()
+    } catch (e: Exception) {
+      e.printStackTrace()
```
Ditto

Sarthak2601 replied:
Done.

rt4914 replied:
Resolving this now.

vinitamurthi commented on domain/src/main/java/org/oppia/domain/oppialogger/OppiaLogUploader.kt:
```diff
@@ -0,0 +1,66 @@
+private const val OPPIA_EXCEPTION_WORK = "OPPIA_EXCEPTION_WORK_REQUEST"
+private const val OPPIA_EVENT_WORK = "OPPIA_EVENT_WORK_REQUEST"
+
+class OppiaLogUploader() {
```
I am a bit unsure what the purpose of this class is. This is not an uploader since its not doing any uploading. It looks like its only creating the worker requests right? Can we rename this class accordingly?

Sarthak2601 replied:
Aah, I named it so because the work requests are after all the ones that will start the process of uploading. Changed it to `OppiaLogUploadWorkRequest`, is that fine now ?

rt4914 replied:
Keeping this open, for further communication in future.

vinitamurthi reviewed (changes requested):
I think you can add tests now

Sarthak2601 committed:
fixes.

Sarthak2601 committed:
app-module test fix.

Sarthak2601 committed:
Merge branch 'develop' into oppia-work-manager

# Conflicts:
#	app/src/main/java/org/oppia/app/application/OppiaApplication.kt

Sarthak2601 committed:
inprogress

Sarthak2601 committed:
inprogress

Sarthak2601 committed:
tests in progress.

Sarthak2601 committed:
Worker Injection setup.

Sarthak2601 committed:
Merge branch 'develop' into oppia-work-manager

# Conflicts:
#	app/src/main/java/org/oppia/app/application/OppiaApplication.kt

Sarthak2601 committed:
worker factory setup.

Sarthak2601 committed:
worker tests.

Sarthak2601 committed:
merging of request and worker class.

Sarthak2601 committed:
work request invocation on app creation setup.

Sarthak2601 committed:
tests.

Sarthak2601 committed:
app-module tests fix.

Sarthak2601 committed:
lint fix.

Sarthak2601 committed:
comments.

BenHenning commented:
Haven't looked at this in depth yet, but @Sarthak2601 I believe you'll need to add new dependencies to BUILD.bazel files as needed to fix the broken Bazel build. In general, if you need to change a gradle file you'll almost certainly also need to change the corresponding Bazel file for that module. See also: https://docs.google.com/document/d/1V0RMehfi3GLVsZ6dyin1nTo4_TO28LACp4wPlEnvnM0/edit#bookmark=id.lt3lnonhm4zo.

Sarthak2601 committed:
bazel building.

BenHenning commented on app/src/main/java/org/oppia/app/application/OppiaApplication.kt:
```diff
@@ -30,4 +32,8 @@ class OppiaApplication :
     component.getApplicationStartupListeners().forEach(ApplicationStartupListener::onCreate)
   }
+
+  override fun getWorkManagerConfiguration(): Configuration {
```
This should be done via modules not directly in components.

Sarthak2601 replied:
Done.

Sarthak2601 replied:
I tried implementing this via the module and here's what I did - 
Implemented the `Configuration.Provider` in the WorkManagerInitialiser and then initialised the work manager in its on create and then bound them in the module. I had to remove the default work manager initialisation using a content provider in the manifest. All of this worked perfectly in production but didn't seem to work in tests as the work manager was getting initialised again and again. 

So I changed the implementation to just providing the configuration via module and then returning that configuration in the `getWorkManagerConfiguration()`. This way production worked, tests worked and we didn't need to alter the default initialisation of the work manager. Is this fine ?

BenHenning replied:
I'm slightly confused. I actually would expect the work manager to be initialized repeatedly in tests since each test recreates the test application.

With this solution, don't we need to update all test applications to also be a Configuration.Provider? How does it actually get hooked up?

Sarthak2601 replied:
I also expected the same but once it's initialised, it gives error when we re-initialise it again. Also, it got hooked up to the tests as is. Whenever we want to use their functionality, we can initialise the testWorkManager via the api provided by the testing dependency.

BenHenning replied:
Looks like this is resolved now.

BenHenning commented on domain/build.gradle:
```diff
@@ -68,6 +70,9 @@ dependencies {
       'org.robolectric:robolectric:4.3',
       project(":testing"),
   )
+  androidTestImplementation(
```
We don’t use instrumentation tests for domain layer tests.

Sarthak2601 replied:
Removed.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkRequest.kt:
```diff
@@ -0,0 +1,80 @@
+  private val workerConstraints = Constraints.Builder()
+    .setRequiredNetworkType(NetworkType.CONNECTED)
+    .setRequiresBatteryNotLow(true)
+    .setRequiresStorageNotLow(true)
```
Why this constraint? Won’t this prevent uploading logs in cases when storage is low and could be freed by uploading logs?

Sarthak2601 replied:
Okay, done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkRequest.kt:
```diff
@@ -0,0 +1,80 @@
+  private val context: Context
+) : ApplicationStartupListener {
+
+  private val workerConstraints = Constraints.Builder()
```
Please add tests to ensure these constraints don’t change accidentally.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,112 @@
+const val LOG_UPLOAD_WORKER = "log_upload_worker"
+
+/** [Worker] class that extracts log reports from the cache store and logs them to the remote service. */
+@Singleton
```
I wouldn’t expect to be using a singleton injectable class with a factory pattern. Shouldn’t the factory return a new instance for each call?

Sarthak2601 replied:
Yes, you're right. Modified.

BenHenning replied:
I see that you removed the scope, but the class is still injectable. As mentioned, constructed classes do not need injection constructors.

Also, this constructor should be private since it's only called from its factory.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkRequest.kt:
```diff
@@ -0,0 +1,80 @@
+    val workerCase =
+      Data.Builder()
+        .putString(
+          LogUploadWorker.WORKER_CASE_KEY,
```
What purpose does this enum to string conversion serve? In general that’s not a great practice since we can’t rely on enum constants to be a particular value (at least in Java).

Sarthak2601 replied:
Okay so the input data for the worker works as a bundle. It only deals in int, string, float, double, etc. values. Now I need  to handle two cases(event and exception) inside the doWork() of the worker. So to handle both these things I'm doing this way. Should I go for something else?

Sarthak2601 replied:
Modified, PTAL.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  /** Extracts exception logs from the cache store and logs them to the remote service. */
+  private fun exceptionWork(): Result {
```
Name: method names should be verbs or verb phrases. Please revise. Ditto below.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,112 @@
+  private fun eventWork(): Result {
+    return try {
+      val eventStore = analyticsController.getEventLogs()
+      val eventLogs = eventStore.value?.getOrThrow()?.eventLogList
```
Is this being called on the main thread? If not, we shouldn’t be interacting directly with LiveData like this.

Sarthak2601 replied:
Okay, was the way of interaction correct in case of the exceptions ? I've modified this to make the two implementations similar. PTAL.

BenHenning replied:
Sorry, this wasn't a comment about consistency. If this function is not being called on the main thread, we cannot retrieve LiveData's value: it's designed to be used only on the main thread. How does AndroidX's WorkManager behave when doWork() is called?

Sarthak2601 replied:
Changed the implementation to a coroutine worker and handled cache retrieval via the original controller. PTAL.

BenHenning replied:
Looks good--thanks!

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerModule.kt:
```diff
@@ -0,0 +1,36 @@
+
+  @Binds
+  @WorkerKey(LogUploadWorker::class)
+  @IntoMap
```
Why use a map if it only has one element?

Sarthak2601 replied:
I did this because in the future we might want to add other workers as well. Like in Sajal's original project, there was a situation where he had to sync the profiles to the Oppia backend.

BenHenning replied:
I see. I actually suggest not generalizing this right now because it's not clear if the constraints the system is currently run under will also work for other applications of WorkManager. Generalizing may take a bit more thought based on the exact needs of the different workers, and possibly more abstraction since the current solution is heavily dependent on WorkManager, making any future work to be able to switch to another delivery work manager API difficult.

Sarthak2601 replied:
Okay, done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkRequest.kt:
```diff
@@ -0,0 +1,80 @@
+
+/** Enqueues unique periodic work requests for uploading events and exceptions to the remote service on application creation. */
+@Singleton
+class LogUploadWorkRequest @Inject constructor(
```
Suggest: LogUploadWorkManagerInitializer

The current name confused me slightly—this class isn’t representing a request.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkRequestTest.kt:
```diff
@@ -0,0 +1,210 @@
+  }
+
+  @Test
+  fun testWorkRequest_logExcepton_withoutNetwork_callOnCreate_verifyLogToRemoteService() {
```
This and the above test can’t happen in prod: onCreate will always be called before we can actually log anything. Please update both tests to match.

Sarthak2601 replied:
But then there will be previous logs right? Like across different app sessions.

BenHenning replied:
That's a good point. In that case, we should fully simulate the situation & the test names should reflect it. To simulate app restart, you need to recreate the test component since the actual objects will be different in the new application.

Please update this & other tests that are testing "pre-app" behavior.

Sarthak2601 replied:
Due to work manager testing issue, couldn't apply these test cases. However, tests verifying the call of this function and its contents have been put. PTAL.

BenHenning replied:
This seems reasonable--thanks!

BenHenning reviewed (changes requested):
Thanks @Sarthak2601. Took a first pass.

BenHenning commented:
Also, I didn’t see any Bazel files updated, just the workspace. Note that the workspace change enables using a dependency, but doesn’t actually use that dependency. You’ll need to update the domain/BUILD.bazel file.

Sarthak2601 committed:
fixes.

Sarthak2601 committed:
fixes.

Sarthak2601 committed:
bazel.

Sarthak2601 commented:
> Also, I didn’t see any Bazel files updated, just the workspace. Note that the workspace change enables using a dependency, but doesn’t actually use that dependency. You’ll need to update the domain/BUILD.bazel file.

Yep, done.

Sarthak2601 committed:
lint fix.

Sarthak2601 committed:
more lint fix.

Sarthak2601 committed:
app module test + config impl change.

Sarthak2601 committed:
lint fix.

Sarthak2601 committed:
more lint fix.

vinitamurthi commented:
I'm approving just to unblock the PR but please wait for Rajat and Ben's reviews

Sarthak2601 commented:
> I'm approving just to unblock the PR but please wait for Rajat and Ben's reviews

Yes, thanks :)

rt4914 commented on domain/BUILD.bazel:
```diff
@@ -34,6 +35,7 @@ TEST_DEPS = [
         artifact("com.google.truth:truth"),
         artifact("org.jetbrains.kotlin:kotlin-test-junit"),
         artifact("org.jetbrains.kotlin:kotlin-reflect"),
+        artifact("androidx.work:work-testing:2.4.0"),
```
Nit: I think we can follow the alphabetical order here too.

Sarthak2601 replied:
Done.

rt4914 replied:
Resolving this now

rt4914 commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,110 @@
+import org.oppia.util.logging.ExceptionLogger
+import javax.inject.Inject
+
+const val LOG_UPLOAD_WORKER = "log_upload_worker"
```
Suggestion: 
`const val TAG = "LogUploadWorker.tag"`

Shift this to companion object and use it like this:
`LogUploadWorker.TAG ` everywhere outside this file.

Sarthak2601 replied:
I don't actually need it anywhere outside this file as of now but moved it to companion object as we may use it in the future while logging work requests.

rt4914 replied:
Okat

rt4914 reviewed (approved):
LGTM, just a couple of nits.

Sarthak2601 committed:
nits.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/analytics/AnalyticsController.kt:
```diff
@@ -154,11 +154,8 @@ class AnalyticsController @Inject constructor(
-   */
-  fun getEventLogs(): LiveData<AsyncResult<OppiaEventLogs>> {
-    return dataProviders.convertToLiveData(eventLogStore)
+  /** Returns [eventLogStore] which can be later converted to [LiveData] to extract log reports from the store. */
```
```suggestion
  /** Returns a data provider for log reports that have been recorded for upload. */
```

Previous wording caused some confusion for me:
1) Readers don't know what eventLogStore is--that's an implementation detail
2) It's implied DataProviders can be turned into LiveData since that's more or less its purpose

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/LogUploadWorkerTest.kt:
```diff
@@ -0,0 +1,223 @@
+
+@RunWith(AndroidJUnit4::class)
+@Config(manifest = Config.NONE)
+class LogUploadWorkerTest {
```
Here & elsewhere: rename tests & suites to match their production counterparts.

Sarthak2601 replied:
Modified, PTAL.

BenHenning reviewed (changes requested):
Thanks @Sarthak2601! The Bazel changes LGTM for codeowners, but I have some other concerns about the current implementation. PTAL.

BenHenning commented:
Also from our Hangouts chat where you mentioned doWork() is executed on a background thread: have you tested this manually with the actual app yet? I'm fairly sure this won't work as-is because:
1) In doWork() you're not blocking on the main thread which means that LiveData won't ever have the values from the data provider since that takes at least 2 thread hops
2) In doWork() you aren't observing the LiveData, so it never becomes active. Our LiveData bridge only calls into the data provider if it's active to prevent unnecessary work.

Sarthak2601 committed:
Coroutine worker | no map.

Sarthak2601 committed:
changes.

Sarthak2601 committed:
work manager impl correction.

Sarthak2601 committed:
testing for initialiser.

Sarthak2601 committed:
Merge branch 'develop' into oppia-work-manager

# Conflicts:
#	app/src/main/java/org/oppia/app/application/ApplicationComponent.kt
#	app/src/sharedTest/java/org/oppia/app/splash/SplashActivityTest.kt
#	app/src/test/java/org/oppia/app/home/HomeActivityLocalTest.kt
#	app/src/test/java/org/oppia/app/player/exploration/ExplorationActivityLocalTest.kt
#	app/src/test/java/org/oppia/app/player/state/StateFragmentLocalTest.kt
#	app/src/test/java/org/oppia/app/profile/ProfileChooserFragmentLocalTest.kt
#	app/src/test/java/org/oppia/app/story/StoryActivityLocalTest.kt
#	app/src/test/java/org/oppia/app/testing/player/state/StateFragmentAccessibilityTest.kt
#	app/src/test/java/org/oppia/app/topic/info/TopicInfoFragmentLocalTest.kt
#	app/src/test/java/org/oppia/app/topic/lessons/TopicLessonsFragmentLocalTest.kt
#	app/src/test/java/org/oppia/app/topic/questionplayer/QuestionPlayerActivityLocalTest.kt
#	app/src/test/java/org/oppia/app/topic/revisioncard/RevisionCardActivityLocalTest.kt

Sarthak2601 committed:
Test fixes.

Sarthak2601 committed:
lint fixes.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,105 @@
+      eventLogs.let {
+        for (eventLog in it) {
+          eventLogger.logEvent(eventLog)
+          analyticsController.removeFirstEventLogFromStore()
```
This seems a bit odd to me--why don't we just ask the controller to remove all of the events that were successfully logged? Ditto for removeFirstExceptionLogFromStore above.

Sarthak2601 replied:
I've basically done a one by one transaction here so that none of the logs get misplaced in the interim. This seemed reasonable to me. We can keep track of exceptions in a better way here I think.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/exceptions/ExceptionsController.kt:
```diff
@@ -159,13 +158,30 @@ class ExceptionsController @Inject constructor(
+
+  /** Removes the first [ExceptionLog] from the [exceptionLogStore]. */
+  fun removeFirstExceptionLogFromStore() {
+    exceptionLogStore.storeDataAsync(true) { oppiaExceptionLogs ->
```
For this & AnalyticsController, please qualify the ``true`` by using a named argument to provide context to readers.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/analytics/AnalyticsController.kt:
```diff
@@ -154,11 +151,28 @@ class AnalyticsController @Inject constructor(
+    return eventLogStore
+  }
+
+  /** Returns a list of [EventLog] after reading from [eventLogStore]. */
```
Readers don't know what ``eventLogStore`` is. Try phrasing this based on the public API, not the implementation. Ditto for ``ExceptionsController``.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploaderModule.kt:
```diff
@@ -0,0 +1,11 @@
+
+/** Provides Log Uploader related dependencies. */
+@Module
+interface LogUploaderModule {
```
Perhaps: FirebaseLogUploaderModule

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploaderModule.kt:
```diff
@@ -0,0 +1,11 @@
+package org.oppia.domain.oppialogger.loguploader
```
Suggest putting this with the other Firebase impls to keep them together.

Sarthak2601 replied:
Won't that mean adding additional dependencies (work manager) in the utility module of both gradle and bazel just for the sake of this file ?

Sarthak2601 replied:
Done.

BenHenning replied:
You're correct, but one advantage of keeping all "Firebase" things together is that it lets us easily switch out implementations in the future (if we ever moved off Firebase).

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/FirebaseLogUploader.kt:
```diff
@@ -0,0 +1,39 @@
+package org.oppia.domain.oppialogger.loguploader
```
Suggest putting this with the other Firebase deps to keep them together.

Sarthak2601 replied:
Ditto as above.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/FirebaseLogUploader.kt:
```diff
@@ -0,0 +1,39 @@
+private const val OPPIA_EXCEPTION_WORK = "OPPIA_EXCEPTION_WORK_REQUEST"
+
+/** Enqueues work requests for uploading stored event/exception logs to the remote service. */
+@Singleton
```
This is overscoping--we don't actually have any state to scope. I suggest removing this.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/FirebaseLogUploader.kt:
```diff
@@ -0,0 +1,39 @@
+@Singleton
+class FirebaseLogUploader @Inject constructor() : LogUploader {
+
+  /** Enqueues a unique periodic work request for uploading exceptions to the remote service. */
```
Documentation doesn't seem needed here & below since it's overriding behavior defined in ``LogUploader``.

Sarthak2601 replied:
Done.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkManagerInitializerTest.kt:
```diff
@@ -0,0 +1,264 @@
+  @Singleton
+  @Component(
+    modules = [
+      TestModule::class,
+      TestLogReportingModule::class,
+      TestLogStorageModule::class,
+      TestDispatcherModule::class,
+      LogUploadWorkerModule::class,
+      TestLogUploaderModule::class
```
```suggestion
      TestModule::class, TestLogReportingModule::class, TestLogStorageModule::class, TestDispatcherModule::class, LogUploadWorkerModule::class, TestLogUploaderModule::class
```

Prefer tightly packing these (my suggested edit probably will need to be slightly broken up, though).

Sarthak2601 replied:
I've done it in a 2-2 configuration.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkManagerInitializerTest.kt:
```diff
@@ -0,0 +1,264 @@
+@Singleton
+class FakeLogUploader @Inject constructor() : LogUploader {
+
+  private val eventRequestIdList = ArrayList<UUID>()
```
```suggestion
  private val eventRequestIdList = mutableListOf<UUID>()
```

Prefer ``mutableListOf``. Ditto below.

Sarthak2601 replied:
Done.

Sarthak2601 replied:
To be clear on this : The enqueuing of work requests doesn't depend on if we have network connectivity or not. Even if we have internet connection, work requests will be enqueued and the worker will work as is. However, during that time no logs will be saved as all of them will be uploaded directly to the remote service.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,105 @@
+  }
+
+  /** Creates an instance of [LogUploadWorker] by properly injecting dependencies. */
+  class FactoryLogUpload @Inject constructor(
```
Prefer just ``Factory``. Then you can inject ``LogUploadWorker.Factory`` where needed.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,105 @@
+          exceptionsController.removeFirstExceptionLogFromStore()
+        }
+      }
+      Result.success()
```
What if the log operation fails? Should we track it & return a failure in that case, then retry later?

Sarthak2601 replied:
If any of the operation fails, it'll throw an error which will be caught and the result will be failed. All of the un-uploaded logs will then be taken up in the next work manager cycle. Does that make sense?

BenHenning replied:
I don't think that's quite right, unless I'm missing something. The suspend function you're calling on the DataProvider (retrieveData()) won't throw an exception, it'll return an AsyncResult with the failure. Don't we need to check all of these results for potential failures?

Sarthak2601 replied:
Umm..I'm actually using the readDataAsync from the Analytics/Exceptions Controller. Will that not throw an error ? I believe I faced an error while testing.

BenHenning replied:
No, it won't. See the implementation: https://github.com/oppia/oppia-android/blob/2ca2397/data/src/main/java/org/oppia/data/persistence/PersistentCacheStore.kt#L75. Or, at least, it won't under normal operations. Normal failures will be propagated via AsyncResult--that's actually why we have it.

Sarthak2601 replied:
Done.

Sarthak2601 replied:
It seems like there's a misunderstanding here. I am not using retrieveData function anymore as the implementation has been shifted to readDataAsync which returns a deferrable instead of an asyncResult. According to already existing implementations of readDataAsync, we don't need to check for failure cases here. Please correct me if I am wrong here.

rt4914 replied:
Resolving this now.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorker.kt:
```diff
@@ -0,0 +1,105 @@
+      }
+      Result.success()
+    } catch (e: Exception) {
+      consoleLogger.e(TAG, e.toString(), e)
```
Here & above: prefer logging context. Adding the exception will provide sufficient context to the exception, e,g:

```kotlin
consoleLogger.e(TAG, "Failed to upload events", e)
```

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerModule.kt:
```diff
@@ -0,0 +1,26 @@
+  ): LogUploadChildWorkerFactory
+
+  @Binds
+  fun bindWorkerFactory(logUploadWorkerFactory: LogUploadWorkerFactory): WorkerFactory
```
Why is this needed? Can't we just inject the factory where we need to?

Sarthak2601 replied:
Actually that's worker factory that provides context and worker params to the child factory which then injects the other dependencies and clubs all these together to make the worker class. This is done because we can't inject the worker params directly.

BenHenning replied:
I understand why the factory is needed, but I don't quite understand why we can't inject LogUploadWorkerFactory where we're currently injecting WorkerFactory (though I don't actually see this injected anywhere--are you sure this binding is being used?)

Sarthak2601 replied:
Yes, it's getting set up as the default worker factory using the WorkManagerConfiguration at app start up.

BenHenning replied:
That's injecting LogUploadWorkerFactory, not WorkerFactory (which is what this binding is providing). This still seems unnecessary--you should be able to remove it.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerFactory.kt:
```diff
@@ -0,0 +1,22 @@
+  private val workerFactory: LogUploadChildWorkerFactory
+) : WorkerFactory() {
+
+  /** Returns a [LogUploadWorker] after injecting [appContext] and [workerParameters] into [LogUploadChildWorkerFactory]. */
```
Or perhaps: ``Returns a new [LogUploadWorker] for the given context and parameters.``

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerFactory.kt:
```diff
@@ -0,0 +1,22 @@
+  /** Returns a [LogUploadWorker] after injecting [appContext] and [workerParameters] into [LogUploadChildWorkerFactory]. */
+  override fun createWorker(
+    appContext: Context,
+    workerClassName: String,
```
Is it okay to ignore this?

Sarthak2601 replied:
Yes, it's okay. I've verified.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/loguploader/LogUploadChildWorkerFactory.kt:
```diff
@@ -0,0 +1,10 @@
+import androidx.work.WorkerParameters
+
+/** Returns an instance of the worker class after injecting the needed dependencies. */
+interface LogUploadChildWorkerFactory {
```
Why do we need this? Why not just inject LogUploadWorker.Factory directly where needed?

Sarthak2601 replied:
LogUploadWorker.Factory is an implementation of this interface. This interface's instance is then used by the worker factory to inject the context and worker params into.

BenHenning replied:
That part is clear to me, but what isn't clear to me is why can't we inject LogUploadWorker.Factory where we are currently injecting this factory? Why do we need this interface & why can't we use the only implementation of it directly?

Sarthak2601 replied:
The interface is basically to link the appContext and workParams that we receive from the WorkerFactory and relay it in such a form so that we can inject other dependencies as well. 
Note - WorkerFactory gets the `appContext` and `workParams` from the `WorkManagerConfiguration` at application start.

BenHenning replied:
I don't think I'm being really clear here: I'm not suggesting that you stop passing those things. I'm suggesting you delete this interface and in ``LogUploadWorkerFactory`` inject ``LogUploadWorker.Factory``, instead. It still does all of the things you mentioned above, but it avoids needing this class (you should move the documentation to the Factory class, though).

Sarthak2601 replied:
Yes, done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerTest.kt:
```diff
@@ -0,0 +1,245 @@
+  @Singleton
+  @Component(
+    modules = [
+      TestModule::class,
+      TestLogReportingModule::class,
+      TestLogStorageModule::class,
+      TestDispatcherModule::class,
+      LogUploadWorkerModule::class,
+      TestLogUploaderModule::class
```
Ditto for these modules (prefer tightly formatting them).

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerTest.kt:
```diff
@@ -0,0 +1,245 @@
+@Config(manifest = Config.NONE)
+class LogUploadWorkerTest {
+
+  @Rule
+  @JvmField
+  val mockitoRule: MockitoRule = MockitoJUnit.rule()
```
Doesn't seem needed--you're not using captors or mocks. Ditto for LogUploadWorkManagerInitializerTest.

Sarthak2601 replied:
Done.

BenHenning replied:
Doesn't seem removed in the latest changes--am I missing something?

Sarthak2601 replied:
Have you refreshed ? I think it's removed.

BenHenning replied:
I still see this in LogUploadWorkManagerInitializerTest. Please make sure to keep an eye on 'ditto' in comments since it means it applies broadly.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/test/java/org/oppia/domain/oppialogger/loguploader/LogUploadWorkerTest.kt:
```diff
@@ -0,0 +1,245 @@
+  }
+
+  @Test
+  fun testWorker_logException_withoutNetwork_enqueueRequest_verifySuccess() {
```
It seems also worthwhile to test that when there is network that it doesn't enqueue the request (since it should log it directly).

Sarthak2601 replied:
Actually that `withoutNetwork` is basically for putting the exceptions inside the store and the work request will perform its function in presence of network only (in background, when the app instance is closed).

BenHenning replied:
To be clear: I'm actually suggesting that we verify that we _don't_ enqueue the request if we log with internet connectivity on (since it should just log as normal). Am I maybe misunderstanding why we shouldn't do that?

Sarthak2601 replied:
Okay, so work manager will work in the background only. It'll get enqueued at its specified time interval only and will fetch logs from the cache store. Now when the app is running and there is network, it'll not store in the cache store and will log directly itself. If there are no cached log reports in the store, it'll just complete the process of the request. Does that help ?

BenHenning replied:
Yep, I'm actually asking that you verify everything you just said actually works like we expect it to via tests here.

Sarthak2601 replied:
Work manager and event logging are completely separate things here. Request enqueue and dequeue will work regardless of how we are logging our events. The work manager's task is to enqueue a request, fetch the logs from the store and upload them on Firebase. Even if there is network connection, work request will get enqueued to fetch data from the cache store. Logging of an event/exception is a part of the Analytics/Exception controller and not the work manager.

rt4914 replied:
Keeping this open, for further communication in future.

BenHenning reviewed (approved):
Thanks @Sarthak2601! Overall, I think this looks really good. Nearly all of my comments are just nits that I noticed. I think the structure of this PR is great.

BenHenning reviewed (changes requested):
Meant to request changes.

Sarthak2601 committed:
changes.

Sarthak2601 committed:
lint fixes.

Sarthak2601 committed:
dependency changes.

Sarthak2601 committed:
lint fixes.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/analytics/AnalyticsController.kt:
```diff
@@ -156,14 +156,14 @@ class AnalyticsController @Inject constructor(
+  /** Removes the first event log report that had been recorded for upload. */
   fun removeFirstEventLogFromStore() {
-    eventLogStore.storeDataAsync(true) { oppiaEventLogs ->
+    eventLogStore.storeDataAsync(/* updateInMemoryCache= */true) { oppiaEventLogs ->
```
```suggestion
    eventLogStore.storeDataAsync(updateInMemoryCache = true) { oppiaEventLogs ->
```

Optional. Yours is completely acceptable (though you should add a space after the ``/``), though Kotlin gives us this syntax for Kotlin APIs for free & it includes automatic validation.

Sarthak2601 replied:
Done.

BenHenning commented on domain/src/main/java/org/oppia/domain/oppialogger/exceptions/ExceptionsController.kt:
```diff
@@ -163,14 +163,14 @@ class ExceptionsController @Inject constructor(
+  /** Removes the first exception log report that had been recorded for upload. */
   fun removeFirstExceptionLogFromStore() {
-    exceptionLogStore.storeDataAsync(true) { oppiaExceptionLogs ->
+    exceptionLogStore.storeDataAsync(/* updateInMemoryCache = */true) { oppiaExceptionLogs ->
```
Ditto here.

Sarthak2601 replied:
Done.

Sarthak2601 committed:
lint.

BenHenning reviewed (changes requested):
Thanks @Sarthak2601! PTAL at my comments & reply before I approve. Also, I think you'll need to make Bazel changes and that will require my approval per code owners.

Sarthak2601 committed:
minor.

Sarthak2601 commented:
@BenHenning for the code owners, I've added artifacts in Bazel and it's building perfectly. PTAL. Also, I used [this](https://android.jlelse.eu/injecting-into-workers-android-workmanager-and-dagger-948193c17684) article for work manager - dagger integration.

BenHenning commented on utility/BUILD.bazel:
```diff
@@ -23,6 +23,7 @@ kt_android_library(
         "//app:crashlytics_deps",
         artifact("org.jetbrains.kotlinx:kotlinx-coroutines-core"),
         artifact("androidx.appcompat:appcompat"),
+        artifact("androidx.work:work-runtime-ktx:2.4.0"),
```
Remove the version--having it defined in WORKSPACE should be sufficient.

Sarthak2601 replied:
Done.

rt4914 replied:
Resolving this now.

BenHenning commented:
@Sarthak2601 left 1 new comment & replies to most existing threads. PTAL.

BenHenning reviewed (approved):
Approving to unblock, but submission is conditioned on my comment threads being resolved. @rt4914 & @anandwana001 please verify these.

Sarthak2601 committed:
change 1.

Sarthak2601 committed:
child factory removal.

Sarthak2601 committed:
lint fix.

Sarthak2601 committed:
Merge branch 'develop' into oppia-work-manager

BenHenning committed:
Add test applications to all tests which now rely on WorkManager to NOT
be initialized (these tests shouldn't use OppiaApplication anymore).
Includes some fixes for various tests that may have had
flakiness/syncing issues before, and removed one test activity that
wasn't actually used. One test was also ignored because it isn't testing
the correct thing and it fails when the correct assertion is added, so
follow-up work is needed.

Sarthak2601 committed:
comment addition.

rt4914 reviewed (approved):
LGTM, thanks.
